### ISSUE TITLE
[Performance] CatalogSearch Indexer Performance Degradation with Elasticsearch

### ISSUE THREAD

**Issue by pmclain** [magento/magento2#18787]
### Preconditions (*)
1. Magento 2.3-develop#fcf9eed8a2456a88d36cfcca02f6a35ad81d6972
2. Install medium performance profile `magento setup:performance:generate-fixtures setup/performance-toolkit/profiles/ce/medium.xml`
3. Install Elasticsearch 5

### Steps to reproduce (*)
1. Validate MySQL is enabled for search
2. `magento indexer:reindex catalogsearch_fulltext` Using single thread
3. Note index time. Locally I've observed ~4 minutes
4. Configure search to use Elasticsearch5+
5. `magento indexer:reindex catalogsearch_fulltext` Using single thread
6. Note index time. Locally I've observed ~22 minutes

### Expected result (*)
1. Full re-indexing time should be somewhat consistent using different backends.

### Actual result (*)
1. Using Elasticsearch increases re-indexing time by 5x.

**Comment by magento-engcom-team** [comment]
Hi @pmclain. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@pmclain do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by pmclain** [comment]
Part of the slowdown comes from loading the stockItem for each product as the index is built
https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Elasticsearch/Model/Indexer/Plugin/StockedProductsFilterPlugin.php#L70-L93

Loading the stockItems in batches matching the current product batch seems to help. The change below reduced reindexing time by ~20%.

```diff
diff --git a/app/code/Magento/Elasticsearch/Model/Indexer/Plugin/StockedProductsFilterPlugin.php b/app/code/Magento/Elasticsearch/Model/Indexer/Plugin/StockedProductsFilterPlugin.php
index ec18b955a29..4ab3f84dff0 100644
--- a/app/code/Magento/Elasticsearch/Model/Indexer/Plugin/StockedProductsFilterPlugin.php
+++ b/app/code/Magento/Elasticsearch/Model/Indexer/Plugin/StockedProductsFilterPlugin.php
@@ -7,6 +7,7 @@ declare(strict_types=1);
 
 namespace Magento\Elasticsearch\Model\Indexer\Plugin;
 
+use Magento\CatalogInventory\Api\Data\StockItemInterface;
 use Magento\Elasticsearch\Model\Config;
 use Magento\CatalogInventory\Api\StockConfigurationInterface;
 use Magento\CatalogInventory\Api\StockStatusRepositoryInterface;
@@ -39,6 +40,11 @@ class StockedProductsFilterPlugin
      */
     private $stockStatusCriteriaFactory;
 
+    /**
+     * @var StockItemInterface[]
+     */
+    private $stockStatuses = [];
+
     /**
      * @param Config $config
      * @param StockConfigurationInterface $stockConfiguration
@@ -74,15 +80,7 @@ class StockedProductsFilterPlugin
         int $storeId
     ): array {
         if ($this->config->isElasticsearchEnabled() && !$this->stockConfiguration->isShowOutOfStock($storeId)) {
-            $productIds = array_keys($indexData);
-            $stockStatusCriteria = $this->stockStatusCriteriaFactory->create();
-            $stockStatusCriteria->setProductsFilter($productIds);
-            $stockStatusCollection = $this->stockStatusRepository->getList($stockStatusCriteria);
-            $stockStatuses = $stockStatusCollection->getItems();
-            $stockStatuses = array_filter($stockStatuses, function (StockStatusInterface $stockStatus) {
-                return StockStatusInterface::STATUS_IN_STOCK == $stockStatus->getStockStatus();
-            });
-            $indexData = array_intersect_key($indexData, $stockStatuses);
+            $indexData = array_intersect_key($indexData, $this->stockStatuses);
         }
 
         return [
@@ -91,4 +89,53 @@ class StockedProductsFilterPlugin
             $storeId,
         ];
     }
+
+    public function aroundGetSearchableProducts(
+        DataProvider $dataProvider,
+        \Closure $proceed,
+        $storeId,
+        array $staticFields,
+        $productIds = null,
+        $lastProductId = 0,
+        $batch = 100
+    ) {
+        $products = $proceed($storeId, $staticFields, $productIds, $lastProductId, $batch);
+
+        if (!$this->config->isElasticsearchEnabled() && $this->stockConfiguration->isShowOutOfStock($storeId)) {
+            return $products;
+        }
+
+        $productIds = array_column($products, 'entity_id');
+        $relatedProducts = [];
+        foreach ($products as $productData) {
+            $relatedProducts[$productData['entity_id']] = $dataProvider->getProductChildIds(
+                $productData['entity_id'],
+                $productData['type_id']
+            );
+        }
+
+        $productIds = array_merge($productIds, array_values(array_filter($relatedProducts)));
+
+        $this->clearStockStatuses();
+        $this->loadStockStatuses($productIds);
+
+        return $products;
+    }
+
+    private function loadStockStatuses(array $productIds)
+    {
+        $stockStatusCriteria = $this->stockStatusCriteriaFactory->create();
+        $stockStatusCriteria->setProductsFilter($productIds);
+        $stockStatusCollection = $this->stockStatusRepository->getList($stockStatusCriteria);
+        $stockStatuses = array_filter($stockStatusCollection->getItems(), function (StockStatusInterface $stockStatus) {
+            return StockStatusInterface::STATUS_IN_STOCK == $stockStatus->getStockStatus();
+        });
+
+        $this->stockStatuses = $stockStatuses;
+    }
+
+    private function clearStockStatuses()
+    {
+        $this->stockStatuses = [];
+    }
```

**Comment by pmclain** [comment]
`\Magento\Elasticsearch\Model\Adapter\BatchDataMapper\ProductDataMapper::getValuesLabels` is another bottleneck. This function creates a lot of duplicate effort loading attribute options. The hack below reduces index time by another ~20%, but at the cost of memory.

```diff
diff --git a/app/code/Magento/Elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php b/app/code/Magento/Elasticsearch/Model/Adapter/BatchDataMapper/ProductDataM
apper.php
index e4f5de46c4c..76f6d0bfd24 100644
--- a/app/code/Magento/Elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php
+++ b/app/code/Magento/Elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php
@@ -271,8 +271,14 @@ class ProductDataMapper implements BatchDataMapperInterface
      */
     private function getValuesLabels(Attribute $attribute, array $attributeValues): array
     {
+        static $attributeOptions = [];
+
         $attributeLabels = [];
-        foreach ($attribute->getOptions() as $option) {
+        if (!isset($attributeOptions[$attribute->getId()])) {
+            $attributeOptions[$attribute->getId()] = $attribute->getOptions();
+        }
+
+        foreach ($attributeOptions[$attribute->getId()] as $option) {
             if (\in_array($option->getValue(), $attributeValues)) {
                 $attributeLabels[] = $option->getLabel();
             }
```

**Comment by subbu1993** [comment]
@magento-engcom-team Please give me 2.1.15 instance

**Comment by ghost** [comment]
Hi @pmclain thank you for you report, Would you like to create a Pull request for this ?

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-backlog-nazar 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-99208` were created

**Issue Available**: @engcom-backlog-nazar, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by pmclain** [comment]
@magento-engcom-team give me 2.3.1 instance

**Comment by magento-engcom-team** [comment]
Hi @pmclain. Thank you for your request. I'm working on Magento 2.3.1 instance for you

**Comment by magento-engcom-team** [comment]
Hi @pmclain, here is your [Magento instance](https://i-18787-2-3-1.instances.magento-community.engineering).
Admin access: [https://i-18787-2-3-1.instances.magento-community.engineering/admin](https://i-18787-2-3-1.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by pmclain** [comment]
@magento-engcom-team give me 2.3.1 instance

**Comment by magento-engcom-team** [comment]
Hi @pmclain. Thank you for your request. I'm working on Magento 2.3.1 instance for you

**Comment by magento-engcom-team** [comment]
Hi @pmclain, here is your [Magento instance](https://i-18787-2-3-1.instances.magento-community.engineering).
Admin access: [https://i-18787-2-3-1.instances.magento-community.engineering/admin](https://i-18787-2-3-1.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by pmclain** [comment]
@magento give me 2.3.2 instance

**Comment by magento-engcom-team** [comment]
Hi @pmclain. Thank you for your request. I'm working on Magento 2.3.2 instance for you

**Comment by magento-engcom-team** [comment]
Hi @pmclain, here is your [Magento instance](https://i-18787-2-3-2.instances.magento-community.engineering).
Admin access: [https://i-18787-2-3-2.instances.magento-community.engineering/admin](https://i-18787-2-3-2.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by pmclain** [comment]
@magento give me 2.3.3 instance

**Comment by magento-engcom-team** [comment]
Hi @pmclain. Thank you for your request. I'm working on Magento 2.3.3 instance for you

**Comment by magento-engcom-team** [comment]
Hi @pmclain, here is your [Magento instance](https://i-18787-2-3-3.instances.magento-community.engineering).
Admin access: [https://i-18787-2-3-3.instances.magento-community.engineering/admin](https://i-18787-2-3-3.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by pemann** [comment]
Has this been fixed?

**Comment by magento-engcom-team** [comment]
Hi @pmclain. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-24396` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.

**Comment by ioweb-gr** [comment]
@magento-engcom-team Hi, could you let us know the PR that fixed this?
